---
title: 'Lesson 2: Manipulating data frames'
output: html_document
---

> ## Overview  
><br />
Teaching: x mins  
Exercises: 0 mins    
>
**Objectives**:  
>
> 1) Import data into R
> 2) Perform basic data manipulation using Tidyverse package
> 3) Write readable R code
>
><br />

## What is a data frame?
A data frame is the most common way of of storing data in R. Technically, 
it comprises of a list of equal-length vectors. Each element of the list can be 
thought of as a column and the length of each element of the list is the number 
of rows.  

Most biological data are structured in the same manner as a data frame, i.e.
a table that contain one or more column variables and an infinite number of rows 
listing the values of each variable. A data table can be saved as one of these 
file types:

* generic text file: .txt
* tab-separated values file: .tsv
* comma-separated values file: .csv
* excel file: .xls or .xlsx.

## Importing data into R as data frame
Converting tabular data into an R data frame is a natural process. In this section,
we will learn how to import data tables from your local computer into R.  

Create a new RMarkdown file and save it on your local computer. Give the document
an appropriate **title** in the metadata section and replace the default content 
of the document with a header `## Import data` as in Figure 1.
<br />
```{r RStudio_Rmarkdown_chunk, fig.cap= "Figure 1: Fresh new RMarkdown document", echo =FALSE, fig.align='center', out.width="80%"}
#knitr::include_graphics("./RStudio_RMarkdown_chunk.PNG", error = FALSE)
```
<br /><br />

We begin with retrieving a sample data from the cloud. This file contain 
a list of alternatively-spliced exons and its inclusion levels in two biological
samples; Sfpq mutants and siblings.  The data is saved in a tab-delimited 
file format (tsv). To download this file, create a new R
chunk (click ![](./Chunk_insert.PNG)) and type the following code into it:  
```{r include=F, message=FALSE}
tmp <- tempfile()
download.file("https://bit.ly/36J7lx4", tmp)
dat <- read.delim(tmp, stringsAsFactors = T)
```
```{r eval=F}
download.file("https://bit.ly/36J7lx4", "splicing_data.tsv")
```
<br />

The function `download.file` will retrieve the content from the input hyperlink 
and save it to your current working directory under the filename 
`splicing_data.tsv`. To import this tsv into R, we will run the `read.delim` 
function and provide the name of the file as its input. Add the following code 
in the chunk above:
```{r eval=F}
dat <- read.delim("splicing_data.tsv", stringsAsFactors = T)
```
<br />

and run the chunk by pressing the **Run** button (![](./green-arrow.PNG)). Our
data is successfully imported into R and assigned to the variable name `dat`.

> ## Tip
> <br />
> In this example, the data file is found in the current working directory, hence
providing the name of the file as input to `read.delim` is sufficient. However,
to import files not found in the working directory, you would need to provide
the full path to the file as input. To get this information, refer to
[this link for Windows user](https://www.howtogeek.com/670447/how-to-copy-the-full-path-of-a-file-on-windows-10/) 
or [this link for Mac user](https://macpaw.com/how-to/get-file-path-mac).
>
><br />

## Previewing data frames
Once imported, R data frames can be previewed in many different ways.  

The first 6 lines or 6 six rows of the data frame `dat` can be printed using the
`head` or `tail` function. Create a sub-section in your RMarkdown document and
insert these code into a new chunk:
```{r, eval = F}
head(dat)
tail(dat)
```

Upon running this chunk, you should get two tables that look like this:
```{r echo = F}
knitr::kable(head(dat))
knitr::kable(tail(dat))
```

The data frame contain 7 columns, and 159977 rows. The column names are shown
at the very top of the table and each row represent one alternative exon. The 
description of the columns are as such:

|Column name|Description|
|:---|:-----------------|
|Gene_id|Given ID of the parental gene|
|Gene_name|Gene name of the parental gene|
|Coord|Genomic position of exon|
|Strand|Whether exon is in the `+` or `-` strand|
|Type|Type of alternative exon<br>CE: Casette exon<br>AF: Alternative first<br>AL:Alternative last<br>AD: Alternative donor<br>AA: Alternative acceptor<br>RI: Retained intron|
|Psi_control|Percent spliced in index of exon in control|
|Psi_mutant|Percent spliced in index of exon in mutant|

The first 6 rows of the data frame can also be shown in a compact way using `str`
function. Type the following code in a new code chunk and run it:
```{r eval = F}
str(dat)
```

The above code should output:
```{r echo = F}
summary(dat)
```

To summarize the contents of each column, we can run the `summary` function. 
Create a new code chunk and run the following code:
```{r eval = F}
str(dat)
```

which should look like this:
```{r echo = F}
summary(dat)
```

## Basic manipulation 1: Filter

## Basic manipulation 2: Select

## Basic manipulation 3: Mutate

## The pipe `%>%`




> ## Practice on your own
> <br />
> The Markdown syntax contain various formattings that enhances the body 
> of your document. Read this 
[CheatSheet](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)
and attempt the following tasks in a new RMarkdown document.
>
> 1) Create an **ordered list** containing 3 of your favorite food
> 2) Bold the #1 fav from the above list
> 3) Insert an image of your #1 fav food. Can be from a local image on your computer or from a hyperlink.
>
> Render the created RMarkdown as HTML and compare your output with mine ([found here](https://bit.ly/3cFtYX3)).
>
><br />


<br /><br />
<br /><br />
<br /><br />
















